src/v/datalake/coordinator/state_machine.cc

```diff
@@ -10,15 +10,68 @@
 
 #include "datalake/coordinator/state_machine.h"
 
+#include "datalake/coordinator/state.h"
+#include "datalake/coordinator/state_update.h"
 #include "datalake/logger.h"
+#include "model/record_batch_types.h"
 
 namespace datalake::coordinator {
 
+namespace {
+void maybe_log_update_error(
+  update_key key,
+  model::offset o,
+  const checked<std::nullopt_t, stm_update_error>& r) {
+    if (r.has_value()) {
+        return;
+    }
+    // NOTE: inability to update the STM is not necessarily a bug! It indicates
+    // that this update's construction raced with another update that broke an
+    // invariant required to apply update. Expectation is that this update's
+    // caller constructs a new update and tries again if needed.
+    vlog(
+      datalake_log.debug,
+      "Coordinator STM {} update at offset {} didn't apply: {}",
+      key,
+      o,
+      r.error());
+}
+} // namespace
+
 coordinator_stm::coordinator_stm(ss::logger& logger, raft::consensus* raft)
   : raft::persisted_stm<>("datalake_coordinator_stm.snapshot", logger, raft) {}
 
-ss::future<> coordinator_stm::do_apply(const model::record_batch&) {
-    co_return;
+ss::future<> coordinator_stm::do_apply(const model::record_batch& b) {
+    if (b.header().type != model::record_batch_type::datalake_coordinator) {
+        return ss::now();
+    }
+    auto iter = model::record_batch_iterator::create(b);
+    while (iter.has_next()) {
```

Comment:
nit: for_each_record_async()  if possible (since the loop also serdes stuff).

Reply:
hm, is this different from just adding co_await maybe_yield() in the loop? While loops feel easier on the brain than for_each+functions

Reply:
think you could just use async variants of serde::read_async() and avoid a yield() call.

Reply:
Done